File: cloudflare/control-plane/src/replacement.ts
import { Json, requireThat, State } from "./shared";

const RESET_R2_PAGES = 4;
const PREVIEW_R2_PAGES = 10;
const TABLES = [
  "sequence", "records", "resources", "staged_fact_rows", "fact_rows",
  "fact_schema", "staged_fact_items", "staged_fact_generations",
  "validated_fact_graphs", "artifact_manifests", "artifact_cleanup",
  "artifact_upload_claims",
] as const;

export function replacementPrefix(workspaceId: string): string {
  return `workspaces/${workspaceId}/artifacts/`;
}

/** Describe the exact workspace-local SQL and R2 scope without mutating it. */
export async function previewWorkspaceReplacement(
  state: State, env: Env, workspaceId: string, exportSha256: string,
): Promise<Json> {
  const tables: Record<string, number> = {};
  for (const table of TABLES) {
    tables[table] = state.sql.exec<{ count: number }>(
      `SELECT COUNT(*) AS count FROM ${table}`,
    ).one().count;
  }
  const recordKinds = Object.fromEntries(state.sql.exec<{ kind: string; count: number }>(
    "SELECT kind,COUNT(*) AS count FROM records GROUP BY kind ORDER BY kind",
  ).toArray().map(row => [row.kind, row.count]));
  const objects = await inspectPrefix(env.ARTIFACTS, replacementPrefix(workspaceId), PREVIEW_R2_PAGES);
  return {
    mode: "preview", workspace_id: workspaceId, expected_export_sha256: exportSha256,
    sql: { delete_all: true, tables, record_kinds: recordKinds },
    r2: objects,
    preserved: ["CT_PRINCIPALS", "CT_CURSOR_KEY", "WORKSPACES binding",
      "ARTIFACTS binding", "other workspace Durable Objects", "other R2 prefixes"],
    quota: "reset does not restore Cloudflare daily counters",
  };
}

/** Delete only one workspace's R2 prefix in bounded, restartable batches. */
export async function deleteWorkspaceArtifactPrefix(
  env: Env, workspaceId: string,
): Promise<{ prefix: string; deleted: number; complete: boolean }> {
  const prefix = replacementPrefix(workspaceId);
  let deleted = 0;
  for (let page = 0; page < RESET_R2_PAGES; page++) {
    // Always restart at the prefix beginning. Deleting a listed page can
    // invalidate a continuation cursor, while this converges safely on retry.
    const objects = await env.ARTIFACTS.list({ prefix, limit: 1000 });
    if (!objects.objects.length) return { prefix, deleted, complete: true };
    const keys = objects.objects.map(object => object.key);
    requireThat(keys.every(key => key.startsWith(prefix)), "replacement_prefix_escape", 500);
    await env.ARTIFACTS.delete(keys);
    deleted += keys.length;
  }
  const remaining = await env.ARTIFACTS.list({ prefix, limit: 1 });
  return { prefix, deleted, complete: remaining.objects.length === 0 };
}

async function inspectPrefix(bucket: R2Bucket, prefix: string, maxPages: number): Promise<Json> {
  let cursor: string | undefined;
  let count = 0;
  let bytes = 0;
  for (let page = 0; page < maxPages; page++) {
    const result = await bucket.list({ prefix, cursor, limit: 1000 });
    count += result.objects.length;
    bytes += result.objects.reduce((total, object) => total + object.size, 0);
    if (!result.truncated || !result.cursor) {
      return { prefix, objects: count, bytes, truncated: false };
    }
    cursor = result.cursor;
  }
  return { prefix, objects: count, bytes, truncated: true };
}

File: cloudflare/control-plane/src/replacement.py
"""Explicit, workspace-scoped authority replacement."""

from __future__ import annotations

from typing import Any

from shared import State, js_get, js_value, require_that, rows

RESET_R2_PAGES = 4
PREVIEW_R2_PAGES = 10
TABLES = (
    "sequence",
    "records",
    "resources",
    "staged_fact_rows",
    "fact_rows",
    "fact_schema",
    "staged_fact_items",
    "staged_fact_generations",
    "validated_fact_graphs",
    "artifact_manifests",
    "artifact_cleanup",
    "artifact_upload_claims",
    "workspace_replacement",
)


def initialize_replacement(state: State) -> None:
    state.sql.exec(
        """CREATE TABLE IF NOT EXISTS workspace_replacement (
        id INTEGER PRIMARY KEY CHECK (id=1), workspace_id TEXT NOT NULL,
        export_sha256 TEXT NOT NULL, status TEXT NOT NULL CHECK(status IN ('incomplete','complete'))
        )"""
    )


def workspace_replacement(state: State) -> dict[str, Any] | None:
    result = rows(
        state.sql.exec(
            "SELECT workspace_id,export_sha256,status FROM workspace_replacement WHERE id=1"
        )
    )
    if not result:
        return None
    row = result[0]
    return {
        key: js_get(row, key) for key in ("workspace_id", "export_sha256", "status")
    }


def mark_workspace_replacement(
    state: State, workspace_id: str, export_sha256: str, status: str
) -> None:
    state.sql.exec(
        """INSERT INTO workspace_replacement VALUES(1,?,?,?)
        ON CONFLICT(id) DO UPDATE SET workspace_id=excluded.workspace_id,
        export_sha256=excluded.export_sha256,status=excluded.status""",
        workspace_id,
        export_sha256,
        status,
    )


def replacement_prefix(workspace_id: str) -> str:
    return f"workspaces/{workspace_id}/artifacts/"


async def inspect_prefix(bucket: Any, prefix: str, max_pages: int) -> dict[str, Any]:
    cursor = None
    count = 0
    byte_count = 0
    for _page in range(max_pages):
        options: dict[str, Any] = {"prefix": prefix, "limit": 1000}
        if cursor:
            options["cursor"] = cursor
        result = await bucket.list(js_value(options))
        objects = rows(js_get(result, "objects", []))
        count += len(objects)
        byte_count += sum(int(js_get(item, "size", 0)) for item in objects)
        if not js_get(result, "truncated", False) or not js_get(result, "cursor"):
            return {
                "prefix": prefix,
                "objects": count,
                "bytes": byte_count,
                "truncated": False,
            }
        cursor = js_get(result, "cursor")
    return {"prefix": prefix, "objects": count, "bytes": byte_count, "truncated": True}


async def preview_workspace_replacement(
    state: State, env: Any, workspace_id: str, export_sha256: str
) -> dict[str, Any]:
    tables: dict[str, int] = {}
    existing = {
        str(js_get(row, "name"))
        for row in rows(
            state.sql.exec("SELECT name FROM sqlite_master WHERE type='table'")
        )
    }
    for table in TABLES:
        if table in existing:
            tables[table] = int(
                js_get(
                    state.sql.exec(f"SELECT COUNT(*) AS count FROM {table}").one(),
                    "count",
                )
            )
    record_kinds = {
        str(js_get(row, "kind")): int(js_get(row, "count"))
        for row in rows(
            state.sql.exec(
                "SELECT kind,COUNT(*) AS count FROM records GROUP BY kind ORDER BY kind"
            )
        )
    }
    objects = await inspect_prefix(
        env.ARTIFACTS, replacement_prefix(workspace_id), PREVIEW_R2_PAGES
    )
    return {
        "mode": "preview",
        "workspace_id": workspace_id,
        "expected_export_sha256": export_sha256,
        "sql": {"delete_all": True, "tables": tables, "record_kinds": record_kinds},
        "replacement": workspace_replacement(state),
        "r2": objects,
        "preserved": [
            "CT_PRINCIPALS",
            "CT_CURSOR_KEY",
            "WORKSPACES binding",
            "ARTIFACTS binding",
            "other workspace Durable Objects",
            "other R2 prefixes",
        ],
        "quota": "reset does not restore Cloudflare daily counters",
    }


async def delete_workspace_artifact_prefix(
    env: Any, workspace_id: str
) -> dict[str, Any]:
    prefix = replacement_prefix(workspace_id)
    deleted = 0
    for _page in range(RESET_R2_PAGES):
        result = await env.ARTIFACTS.list(js_value({"prefix": prefix, "limit": 1000}))
        objects = rows(js_get(result, "objects", []))
        if not objects:
            return {"prefix": prefix, "deleted": deleted, "complete": True}
        keys = [str(js_get(item, "key")) for item in objects]
        require_that(
            all(key.startswith(prefix) for key in keys),
            "replacement_prefix_escape",
            500,
        )
        await env.ARTIFACTS.delete(js_value(keys))
        deleted += len(keys)
    remaining = await env.ARTIFACTS.list(js_value({"prefix": prefix, "limit": 1}))
    return {
        "prefix": prefix,
        "deleted": deleted,
        "complete": len(rows(js_get(remaining, "objects", []))) == 0,
    }
